sql: add LTREE util package

paulniziolek opened the pull request:
#### sql: add ltree util pkg with basic functionality

The newly introduced ltree pkg introduces the basic underlying LTREE internals
for when the LTREE type is to be introduced into CRDB, such as the data
structure, ltree parsing and validation, and accessory methods.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149912)
<!-- Reviewable:end -->

paulniziolek commented on pkg/util/ltree/encoding.go:
```diff
@@ -0,0 +1,83 @@
+//	  bytes byte-form of labels separated by dots
+//
+// See encodings.json for examples.
+func EncodeLTreeBinary(appendTo []byte, l LTree) ([]byte, error) {
+	// To match PG's binary encoding, we also use '1' as a version byte.
+	appendTo = encoding.EncodeUint8Ascending(appendTo, 1)
+	appendTo = encoding.EncodeBytesAscending(appendTo, []byte(l.String()))
+	return appendTo, nil
+}
```
erm, something regarding my LTREE PG Binary encoding/decoding seems off. When I run test cases for binary format of LTREE in `pgwire`, I get "Invalid data row". I will investigate this tmrow.

paulniziolek commented on pkg/sql/pgwire/testdata/encodings.json:
```diff
@@ -856,15 +884,15 @@
 	{
 		"SQL": "'9E+4'::decimal",
 		"Oid": 1700,
-		"Text": "9E+4",
-		"TextAsBinary": [57, 69, 43, 52],
+		"Text": "90000",
+		"TextAsBinary": [57, 48, 48, 48, 48],
```
whoever wrote this probably should've separated this into `generated_encodings.json` and `overwrite_encodings.json`...

mgartner reviewed (changes requested):
I left inline comments. These changes grew larger than what a single commit should contain. I recommend splitting the changes up into several commits, such as:

1. Introduce `ltree` package with basic functionality of the underlying data type and unit tests.
2. Add `tree.DLTree` with basic functionality.
3. Add value-side encoding for Ltrees.
4. Add key-side encoding for Ltrees.

I'd leave out LQUERY and the other unused data type for now. We don't plan to implement them yet and they can easily be added in a separate PR if we want nicer error messages than "unknown type".

> Reviewed 23 of 57 files at r1, 3 of 11 files at r2, 2 of 10 files at r3, 9 of 24 files at r4, 4 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-:-OWS7zvE8GAvflvx100C:b-n22nts)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @golgeek, @rharding6373, and @rytaft)

___
*[`-- commits` line 3 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWRysC85brkitc-FLeR:-OWRysC85brkitc-FLeS:b-lmsmdk):*
nit: remove "Fixes" so it doesn't auto-close that issue.
___
*[`-- commits` line 6 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWRy_raCsuo6db8tdJA:-OWRy_raCsuo6db8tdJB:b-cnjlgb):*
nit: Leave out a release note until the feature is complete. use `Release note: None`. move some of the information in this release note above into the commit message, like explaining exactly what this commit adds, and what is still unsupported.
___
*[`pkg/sql/catalog/colinfo/col_type_info.go` line 118 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS-PvvDgiNoJPqyZdl:-OWS-PvvDgiNoJPqyZdm:b372gf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/catalog/colinfo/col_type_info.go#L118)):*
> ```Go
> 
> 	case types.LQueryFamily, types.LTXTQueryFamily:
> 		// TODO(paulniziolek): LQUERY and LTXTQUERY should be supported as column types in the future
> ```

nit: TODO is unnecessary
___
*[`pkg/util/ltree/label.go` line 17 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS7Wak0ZPLHUqfw58Z:-OWS7Wak0ZPLHUqfw58_:bdj78la) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/util/ltree/label.go#L17)):*
> ```Go
> // Postgres docs mention labels must be less than 256 bytes, but in practice,
> // Postgres has a limit on up to 1000 characters.
> const MaxLabelLength = 1000
> ```

nit: don't export
___
*[`pkg/util/ltree/label.go` line 19 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS7O8n43NWk4E7ra4Q:-OWS7O8n43NWk4E7ra4R:b-cwrf29) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/util/ltree/label.go#L19)):*
> ```Go
> const MaxLabelLength = 1000
> 
> var ErrEmptyLabel = pgerror.New(pgcode.Syntax, "label cannot be empty")
> ```

nit: don't export this if it's not used outside this package.
___
*[`pkg/util/ltree/label.go` line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS7S3gDQvaA2pMsrEV:-OWS7S3gDQvaA2pMsrEW:b-3sav13) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/util/ltree/label.go#L21)):*
> ```Go
> var ErrEmptyLabel = pgerror.New(pgcode.Syntax, "label cannot be empty")
> 
> type Label string
> ```

nit: don't export this type if it's not used outside this package
___
*[`pkg/util/ltree/label.go` line 23 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS7VQkE_nUltz6QtMk:-OWS7VQkE_nUltz6QtMl:bdj78la) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/util/ltree/label.go#L23)):*
> ```Go
> type Label string
> 
> func NewLabel(label string) (Label, error) {
> ```

nit: don't export
___
*[`pkg/cmd/generate-binary/main.go` line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWRzw6P4DWLkHQ_larw:-OWRzw6P4DWLkHQ_larx:bd9p9cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/cmd/generate-binary/main.go#L74)):*
> ```Go
> 	}
> 
> 	// TODO(paulniziolek): We can probably dynamically create this map by the ExtensionTypeName map in oidext.go
> ```

nit missing a period? Why would we want to dynamically create this map? This seems find to me so you might just remove this TODO.
___
*[`pkg/util/ltree/ltree.go` line 23 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS7_qB18NtDx0ywHKf:-OWS7_qB18NtDx0ywHKg:b-wow3yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/util/ltree/ltree.go#L23)):*
> ```Go
> type LTree struct {
> 	// Path is an ordered slice of Label types that make up a path in a LTREE column.
> 	Path []Label
> ```

nit: don't export `Path` to prevent users from mucking with the type's internals. 
___
*[`pkg/util/ltree/ltree.go` line 65 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS7lb68Ywwx4_vmbOO:-OWS7lb68Ywwx4_vmbOP:b-le3m22) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/util/ltree/ltree.go#L65)):*
> ```Go
> 
> // Size returns the size of the LTree in bytes, which is the sum of the label lengths.
> func (l LTree) Size() int {
> ```

nit: `ByteSize`? Should this include separator bytes? I think so if all usages need that.
___
*[`pkg/sql/sem/tree/datum.go` line 4543 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS5kYyEcMWl28qBkCW:-OWS5kYyEcMWl28qBkCX:bxw9bjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/sem/tree/datum.go#L4543)):*
> ```Go
> 		return 0, makeUnsupportedComparisonMessage(d, other)
> 	}
> 	l, r := d.String(), v.String()
> ```

Allocating strings to perform a comparison is expensive. We should be able to perform the comparison without allocating. The underlying `ltree.LTree` type should expose a compare function that we can use here. 
___
*[`pkg/sql/sem/tree/datum.go` line 4558 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS6Bzk9-grLTTQ97ZU:-OWS6Bzk9-grLTTQ97ZV:b40gim9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/sem/tree/datum.go#L4558)):*
> ```Go
> 			return nil, false
> 		}
> 		emptyLTree, _ := ParseDLTree("")
> ```

This looks incorrect. The empty Ltree is not the previous LTree value for all single-character Ltrees.
___
*[`pkg/sql/sem/tree/datum.go` line 4583 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS6SyN-pUx2557edEr:-OWS6SyO0EsQfA78aujs:b-pxwtpo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/sem/tree/datum.go#L4583)):*
> ```Go
> // Min implements the Datum interface.
> func (d *DLTree) Min(ctx context.Context, cmpCtx CompareContext) (Datum, bool) {
> 	return &DLTree{}, false
> ```

nit: `return NewDLTree(ltree.Empty)`
___
*[`pkg/sql/sem/tree/datum.go` line 4597 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS6bG19VB8FJ-ayiGl:-OWS6bG19VB8FJ-ayiGm:b-6s6ymc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/sem/tree/datum.go#L4597)):*
> ```Go
> func (d *DLTree) Format(ctx *FmtCtx) {
> 	buf, f := &ctx.Buffer, ctx.flags
> 	pathStr := d.LTree.String()
> ```

String allocation here will be expensive too. `Format` is probably not in _very_ hot paths like encoding and comparison are, but it might be nice to avoid the allocations from the get-go so we don't have to optimize later. Exposing a format-like method from the ltree package would be good.
___
*[`pkg/sql/randgen/datum.go` line 326 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS0VAp5fgBCaACVJdt:-OWS0VAqCK0gtgAQhi2L:b-d213p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/randgen/datum.go#L326)):*
> ```Go
> 			if length > 1 {
> 				// ltree cannot have empty labels if it has more than one label.
> 				labelLength += 1
> ```

nit: it'd be simpler to special case the empty label and avoid the conditional here. Exporting an empty Ltree value in the ltree package would be nice too


```go
length := rng.Intn(10)
if length == 1 && rng.Intn(4) == 0 {
    return tree.NewDLtree(ltree.Empty)
}
// ...
for i := range labels {
    // Labels cannot be empty.
    labelLen := rng.Intn(9) + 1
    // ...
}
```
___
*[`pkg/sql/randgen/type.go` line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS3BIhDSAxQZu5qB7s:-OWS3BIhDSAxQZu5qB7t:b-l3r8kl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/randgen/type.go#L52)):*
> ```Go
> 			// TODO(#22513): Temporarily don't include Jsonpath
> 		case oidext.T_lquery, oidext.T_ltxtquery:
> 			// TODO(paulniziolek): temporarily don't include ltree querying types
> ```

nit: should capitalize, add period.
___
*[`pkg/sql/rowenc/keyside/encode.go` line 186 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS4fYJB1eNEuT_ZGhA:-OWS4fYJB1eNEuT_ZGhB:b7i851f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/rowenc/keyside/encode.go#L186)):*
> ```Go
> 	case *tree.DLTree:
> 		if dir == encoding.Ascending {
> 			return encoding.EncodeBytesAscending(b, []byte(t.LTree.String())), nil
> ```

This will definitely cause more allocations that we need. Let's sync about to talk about how to eliminate those.
___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 883 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS05aM6lIbo0AIFnYQ:-OWS05aM6lIbo0AIFnYR:b-sc3scx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/pgwire/pgwirebase/encoding.go#L883)):*
> ```Go
> 					"unsupported ltree version %d", version)
> 			}
> 			// we skip over the version byte when parsing binary LTREE.
> ```

nit: "we skip..." => "Skip ..."
___
*[`pkg/sql/rowenc/keyside/decode.go` line 286 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS3xSw4XaoI5SnUnQE:-OWS3xSw4XaoI5SnUnQF:b3ptlfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/rowenc/keyside/decode.go#L286)):*
> ```Go
> 			return nil, nil, err
> 		}
> 		datum, err := tree.ParseDLTree(string(pathBytes))
> ```

Converting from `[]byte` to `string` here might cause an allocation, which would be costly. We should (1) determine if it does, and (2) if so, parse the bytes directly instead of converting to a string. Let's sync up about (1).


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/util/ltree/ltree.go` line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS8nDh6qhZGTfHyXge:-OWS8nDi--NObt1Uro80:b-dkgog6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/util/ltree/ltree.go#L21)):*
> ```Go
> const MaxNumOfLabels = 65535
> 
> type LTree struct {
> ```

`ltree.LTree` is redundant. I'm a fan of using the shortened name `T` for types exported by packages that share the same name. Usage outside the package would look like `ltree.T`. We do something similar for `volatility.V`: https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/sem/volatility/volatility.go#L20


<!-- Sent from Reviewable.io -->

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149912)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @golgeek, @mgartner, @rharding6373, and @rytaft)

___
*[`pkg/sql/rowenc/keyside/decode.go` line 286 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS3xSw4XaoI5SnUnQE:-OWaU-p8DIgXIY916SpZ:b3kxlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/rowenc/keyside/decode.go#L286)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Converting from `[]byte` to `string` here might cause an allocation, which would be costly. We should (1) determine if it does, and (2) if so, parse the bytes directly instead of converting to a string. Let's sync up about (1).
</blockquote></details>

I think this would cause an allocation... We should sync up about this, maybe I can implement something like `ParseDLTreeFromBytes`.  Ditto for valueside as well. 
___
*[`pkg/sql/rowenc/keyside/encode.go` line 186 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS4fYJB1eNEuT_ZGhA:-OWahvcKBwpak6Ronhpg:b-80tlqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/rowenc/keyside/encode.go#L186)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This will definitely cause more allocations that we need. Let's sync about to talk about how to eliminate those.
</blockquote></details>

Sure.
___
*[`pkg/sql/sem/tree/datum.go` line 4543 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS5kYyEcMWl28qBkCW:-OWaeSJA0PhAgxaUW_OQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/sem/tree/datum.go#L4543)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Allocating strings to perform a comparison is expensive. We should be able to perform the comparison without allocating. The underlying `ltree.LTree` type should expose a compare function that we can use here.
</blockquote></details>

Done.
___
*[`pkg/sql/sem/tree/datum.go` line 4558 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS6Bzk9-grLTTQ97ZU:-OWahxzl614uQr_DXkLB:bglr85) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/sem/tree/datum.go#L4558)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This looks incorrect. The empty Ltree is not the previous LTree value for all single-character Ltrees.
</blockquote></details>

Not sure if I understood `Prev` correctly. Should this be lexicographically returning the previous LTree value? (ex: 'A.B' -> 'A.A and 'A.A' -> 'A')? 
___
*[`pkg/sql/sem/tree/datum.go` line 4597 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWS6bG19VB8FJ-ayiGl:-OWafjrC2Ti_pq_WFpzy:bv0w00s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d071566179ca1503e6a47b5b479062ee3d8f6c/pkg/sql/sem/tree/datum.go#L4597)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

String allocation here will be expensive too. `Format` is probably not in _very_ hot paths like encoding and comparison are, but it might be nice to avoid the allocations from the get-go so we don't have to optimize later. Exposing a format-like method from the ltree package would be good.
</blockquote></details>

Done. But this proposed impl only optimizes the path for formatting for bare strings or pgwire. 


<!-- Sent from Reviewable.io -->

paulniziolek committed:
sql: add ltree util pkg with basic functionality

The newly introduced ltree pkg introduces the basic underlying LTREE internals
for when the LTREE type is to be introduced into CRDB, such as the data
structure, ltree parsing and validation, and accessory methods.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

mgartner reviewed (changes requested):
> Reviewed 81 of 81 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149912)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @golgeek, @paulniziolek, @rharding6373, and @rytaft)

___
*[`pkg/util/ltree/ltree_test.go` line 13 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvngMa3EK0wDeNmMMJ:-OWvngMa3EK0wDeNmMMK:b-m3sie6) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree_test.go#L13)):*
> ```Go
> 	tests := []struct {
> 		input   string
> 		want    string
> ```

nit: `expected`
___
*[`pkg/util/ltree/ltree_test.go` line 77 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvnlhK3iAY3fAzHkK4:-OWvnlhK3iAY3fAzHkK5:bskjo5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree_test.go#L77)):*
> ```Go
> 				return
> 			}
> 			require.NoError(t, err)
> ```

```
if _ {
      t.Errorf("expected %q, got: %q\n", tc.expected, got.String())
}
```
___
*[`pkg/util/ltree/ltree.go` line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvkd33BC9f-sJqd0bV:-OWvkd33BC9f-sJqd0bW:b-gxlvh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L28)):*
> ```Go
> type T struct {
> 	// path is an ordered slice of label types that make up a path in a LTREE column.
> 	path []label
> ```

nit: if we use `[]string` here then we avoid allocating both a `[]string` when we `strings.Split` and a `[]label`.
___
*[`pkg/util/ltree/ltree.go` line 54 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvlhEg-JET8yY4GyU0:-OWvlhEg-JET8yY4GyU1:bf6y81w) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L54)):*
> ```Go
> 
> // String returns the string representation of T.
> func (lt T) String() string {
> ```

nit: put this next to `Format`, and use `Format` instead of duplicating logic
___
*[`pkg/util/ltree/ltree.go` line 73 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvlakJ8NmOXn5ysvY_:-OWvlakJ8NmOXn5ysvYa:b47vipo) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L73)):*
> ```Go
> 
> // LabelAt returns the label at the specified index in an LTree path.
> func (lt T) LabelAt(idx int) (string, error) {
> ```

nit: could probably just call this `Label`
___
*[`pkg/util/ltree/ltree.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvoV6m4ZBqkAXaDkow:-OWvoV6m4ZBqkAXaDkox:b-unr3g2) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L91)):*
> ```Go
> 
> // Compare compares two LTrees lexicographically based on their labels.
> func (lt T) Compare(other T) int {
> ```

This should be unit tested.
___
*[`pkg/util/ltree/ltree.go` line 116 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvln47AqJKh8gg3UVW:-OWvln47AqJKh8gg3UVX:bl9205l) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L116)):*
> ```Go
> }
> 
> // ByteSize returns the size of the T in bytes, which is the sum of the label lengths and their path separators.
> ```

nit: does this exceed 80 characters wide? comments should wrap at 80
___
*[`pkg/util/ltree/ltree.go` line 117 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvoXWg841a2MKAKeeP:-OWvoXWg841a2MKAKeeQ:b36au5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L117)):*
> ```Go
> 
> // ByteSize returns the size of the T in bytes, which is the sum of the label lengths and their path separators.
> func (lt T) ByteSize() int {
> ```

Simple tests for ByteSize would be nice too.
___
*[`pkg/util/ltree/ltree.go` line 128 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvm0nzEdYCshzb3PIi:-OWvm0nzEdYCshzb3PIj:b-sqkevb) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L128)):*
> ```Go
> }
> 
> // Copy creates a deep copy of T.
> ```

nit: remove "deep copy"
___
*[`pkg/util/ltree/ltree.go` line 143 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWvn0iu01iRQ-fP3Qal:-OWvn0iu01iRQ-fP3Qam:b-74abhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/961bb9c10ec366add56e4892b442d994f0d37085/pkg/util/ltree/ltree.go#L143)):*
> ```Go
> 	lastLabel := string(lt.path[lt.Len()-1])
> 	if prevLabelStr := prevLabel(lastLabel); prevLabelStr != "" {
> 		newLbl, err := newLabel(prevLabelStr)
> ```

nit: `prevLabel` should only return valid labels, and then you don't need to validate the label and check for validation error.


<!-- Sent from Reviewable.io -->

yuzefovich commented on pkg/sql/sem/builtins/fixed_oids.go:
```diff
@@ -2749,6 +2749,32 @@ var builtinOidsArray = []string{
+	2811: `ltree(text: text) -> ltree`,
+	2812: `ltree(string: string) -> ltree`,
+	2813: `ltree(varchar: varchar) -> ltree`,
+	2814: `ltree(ltree: ltree) -> ltree`,
```
I don't think so - the build fails and it prints out what is missing, so some copy-paste is required, at least for now.
<!-- Reviewable comment -OWwiPxJ6_sej8-G2ZL4:bizz32o -->
<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice work! I have a few minor suggestions.

> Reviewed 1 of 78 files at r7, 79 of 79 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-:-OWwiLArBEKVOgPacSt3:b-tvs61o)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @paulniziolek)

___
*[`pkg/util/ltree/ltree_test.go` line 1 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWwhGQm0po8STN50BeV:-OWwhGQm0po8STN50BeW:bl0scdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/90336c52607f99d3240aba85ed28a9d7aaaf18ee/pkg/util/ltree/ltree_test.go#L1)):*
> ```Go
> package ltree
> ```

nit: missing the license header.
___
*[`pkg/util/ltree/ltree.go` line 6 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWweY1VAH_Ck4jNCjx6:-OWweY1VAH_Ck4jNCjx7:b-avklcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/90336c52607f99d3240aba85ed28a9d7aaaf18ee/pkg/util/ltree/ltree.go#L6)):*
> ```Go
> // included in the /LICENSE file.
> 
> package ltree
> ```

nit: let's mark the new package as owned by Queries in `.CODEOWNERS`.
___
*[`pkg/util/ltree/ltree.go` line 107 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWwfNnkCss8Euqcoqg_:-OWwfNnkCss8Euqcoqga:bmckf89) ([raw file](https://github.com/cockroachdb/cockroach/blob/90336c52607f99d3240aba85ed28a9d7aaaf18ee/pkg/util/ltree/ltree.go#L107)):*
> ```Go
> func (lt T) Compare(other T) int {
> 	minLen := lt.Len()
> 	if other.Len() < minLen {
> ```

nit: Go now has builtin `min` function.
___
*[`pkg/util/ltree/ltree.go` line 199 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWwgGIj3_-avln8AX4m:-OWwgGIj3_-avln8AX4n:b487dx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/90336c52607f99d3240aba85ed28a9d7aaaf18ee/pkg/util/ltree/ltree.go#L199)):*
> ```Go
> 
> // isValidChar returns true if the character is valid in an LTree label.
> func isValidChar(c byte) bool {
> ```

nit: earlier today we talked about not doing premature optimization, but I can't help myself but leave an optimization-related comment here lol. Given that only a small set of characters are valid, we could initialize a `map` that would store all of them, so the check for validity would be a map lookup. (Generally it's good idea to avoid branching with `if` if possible.)

Furthermore, we can store the prev char as the value in the map to use in `prevChar` function (although the `switch` there is small, so I'm not sure whether it would actually be more performant). 

Probably leave a TODO for now to look into this later.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job!

> Reviewed 1 of 78 files at r7, 1 of 79 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-:-OX-fFav2jMljPrGNS2w:b-46j5c8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @paulniziolek and @yuzefovich)

___
*[`pkg/util/ltree/ltree.go` line 199 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OWwgGIj3_-avln8AX4m:-OX-e1sd9fQAka9nDy4B:b-3twybe) ([raw file](https://github.com/cockroachdb/cockroach/blob/90336c52607f99d3240aba85ed28a9d7aaaf18ee/pkg/util/ltree/ltree.go#L199)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: earlier today we talked about not doing premature optimization, but I can't help myself but leave an optimization-related comment here lol. Given that only a small set of characters are valid, we could initialize a `map` that would store all of them, so the check for validity would be a map lookup. (Generally it's good idea to avoid branching with `if` if possible.)

Furthermore, we can store the prev char as the value in the map to use in `prevChar` function (although the `switch` there is small, so I'm not sure whether it would actually be more performant). 

Probably leave a TODO for now to look into this later.
</blockquote></details>

I'd be surprised if a map lookup beats this conditional. Maybe a sparse array would, e.g.:

```go
var prevChar = [...]byte{
	'0': '-',
	'A': '9',
	// ...
}
```
___
*[`pkg/util/ltree/ltree.go` line 99 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OX-OAsA3fXSfefZuPnx:-OX-OAsA3fXSfefZuPny:b-zek4h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7b1a42c48e1471fcb16182de2d846bf919d4341/pkg/util/ltree/ltree.go#L99)):*
> ```Go
> func (lt T) LabelAt(idx int) (string, error) {
> 	if idx < 0 || idx >= lt.Len() {
> 		return "", pgerror.Newf(pgcode.InvalidParameterValue, "index %d out of bounds.", idx)
> ```

nit: remove trailing `.` in string
___
*[`pkg/util/ltree/ltree.go` line 197 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-OX-esaI5gxLDYOrY88R:-OX-esaI5gxLDYOrY88S:b-8y27dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7b1a42c48e1471fcb16182de2d846bf919d4341/pkg/util/ltree/ltree.go#L197)):*
> ```Go
> // isValidChar returns true if the character is valid in an LTree label.
> func isValidChar(c byte) bool {
> 	if (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c == '_') || (c == '-') {
> ```

nit: no need for `if`, just return the value of the conditional


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149912#-:-OX0Cvb431DAajjVtnWC:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @paulniziolek and @yuzefovich)



<!-- Sent from Reviewable.io -->
